### ISSUE TITLE
cmd/compile: panic on go get golang.org/x/sys/unix with go 1.9.x

### ISSUE THREAD

**Issue by nono** [golang/go#25561]
### What version of Go are you using (`go version`)?

`go version go1.9 linux/amd64`

### Does this issue reproduce with the latest release?

No, it works fine with go1.10.

### What operating system and processor architecture are you using (`go env`)?

```
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/travis/gopath"
GORACE=""
GOROOT="/home/travis/.gimme/versions/go1.9.linux.amd64"
GOTOOLDIR="/home/travis/.gimme/versions/go1.9.linux.amd64/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build054042360=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
```

### What did you do?

I cloned github.com/cozy/cozy-stack and run `go get -t -v ./...` from here.

### What did you expect to see?

A success: all dependencies are installed

### What did you see instead?

```
golang.org/x/sys/unix
# golang.org/x/sys/unix
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x97b3ab]
goroutine 1 [running]:
cmd/compile/internal/gc.dumpasmhdr()
	/usr/local/go/src/cmd/compile/internal/gc/export.go:389 +0x2ab
cmd/compile/internal/gc.Main(0xb73f90)
	/usr/local/go/src/cmd/compile/internal/gc/main.go:633 +0x280d
main.main()
	/usr/local/go/src/cmd/compile/main.go:49 +0x95
```

This issue was reproduced on our servers and on our CI (travis). For travis, the log is available: https://travis-ci.org/cozy/cozy-stack/jobs/383555937#L466

**Comment by tklauser** [comment]
Can reproduce this here with go 1.9.6.

I assume this requires a backport of https://golang.org/cl/80759 to the 1.9 branch?

/cc @ianlancetaylor @mdempsky

**Comment by tklauser** [comment]
FWIW the culprit in `x/sys/unix` seems to be this type alias:

```go
type Nfgenmsg = struct {
	Nfgen_family uint8
	Version      uint8
	Res_id       uint16
}
```
from https://github.com/golang/sys/blob/31355384c89b50e6faeffdb36f64a77a8210188e/unix/ztypes_linux_amd64.go#L1405-L1409, introduced by https://golang.org/cl/114518

**Comment by tklauser** [comment]
As noted in #25563, I think this should also be fixed in `x/sys/unix`. Let's make this issue about the fix in cmd/compile for go 1.9 and #25563 about the fix in `x/sys/unix`.

**Comment by ganelon13** [comment]
got the same problem with go1.9.3 linux/amd64

# golang.org/x/sys/unix
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x97ca3b]

goroutine 1 [running]:
cmd/compile/internal/gc.dumpasmhdr()
        /usr/local/go/src/cmd/compile/internal/gc/export.go:389 +0x2ab
cmd/compile/internal/gc.Main(0xb770b8)
        /usr/local/go/src/cmd/compile/internal/gc/main.go:633 +0x280d
main.main()
        /usr/local/go/src/cmd/compile/main.go:49 +0x95

**Comment by jonnenauha** [comment]
Getting this as well with 1.9.4. The package should be reverted or modified to work with 1.8/1.9 if there is no good reason for this (until 1.9 gets a fix). Otherwise we (users) are basically forced to wait for a new 1.9 release or update to 1.10.

I'm going to update now, cant wait for this if build bots can't build any projects with this package imported.

**Comment by tklauser** [comment]
@jonnenauha yes, as mentioned in #25563 the issue in `x/sys/unix` should be fixed regardless of a fix in the 1.9 branch. I'll send one shortly.

**Comment by Deleplace** [comment]
Got the same problem with 1.9.2. Upgraded to 1.10.2, now everything works fine.
